northstar/src/runtime/process/init.rs

```diff
@@ -30,11 +30,130 @@ pub(super) struct Init {
     pub seccomp: Option<AllowList>,
 }
 
+/// This is the type used to communicate with the init process from the runtime.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum InitMessage {
+    NewChild { pid: i32 },
+    DeadChild { pid: i32, exit_status: ExitStatus },
+}
+
+/// Init commands
+#[derive(Debug, Serialize, Deserialize)]
+pub enum InitApi {
+    /// Executes an command
+    Exec {
+        path: CString,
+        args: Vec<CString>,
+        env: Vec<CString>,
+    },
+}
+
+/// Init process main function
+fn init_main(
+    mut rt_tx: channel::Sender<InitMessage>,
+    mut rt_rx: channel::Receiver<InitApi>,
+    seccomp: Option<AllowList>,
+) -> ! {
+    enum Payload {
+        ChildExit { pid: i32, exit_status: ExitStatus },
+        Command(InitApi),
+    }
+    let (tx, rx) = mpsc::channel();
+
+    // This thread forwards the commands coming from the runtime
+    let command_tx = tx.clone();
+    let _command_thread = std::thread::spawn(move || loop {
+        let command = rt_rx
+            .recv()
+            .expect("Failed to receive init command from runtime");
+        command_tx.send(Payload::Command(command)).unwrap();
+    });
+
+    // This thread sends the exit status of the child processes
+    let exit_status_tx = tx;
+    let reaper_thread = std::thread::spawn(move || {
+        loop {
+            match waitpid(None, None) {
+                Ok(WaitStatus::Exited(pid, status)) => {
+                    exit_status_tx
+                        .send(Payload::ChildExit {
+                            pid: pid.as_raw(),
+                            exit_status: ExitStatus::Exit(status),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Signaled(pid, status, _)) => {
+                    exit_status_tx
+                        .send(Payload::ChildExit {
+                            pid: pid.as_raw() as i32,
+                            exit_status: ExitStatus::Signalled(status as u8),
+                        })
+                        .unwrap();
```

Comment:
We have to make sure that the whole process panics upon this unwrap.
When this thread panics, `tx` is dropped.  The other forward thread also owns a copy of tx and we will not notice that the reaper is gone.

Reply:
I just tested adding a `panic!` after the reaper thread is `unparked` and the whole process panics. I think this is just because we use that `panic = "abort"` policy.

Reply:
Yes. Since this crate is a lib the `panic` field in the manifest must be set in the user of the northstar crate. Needs to be added to the docs.